File: Android_GUI/src/com/github/chenxiaolong/dualbootpatcher/switcher/RomDetailAdapter.java
/*
 * Copyright (C) 2015  Andrew Gunnerson <andrewgunnerson@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.github.chenxiaolong.dualbootpatcher.switcher;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.drawable.Drawable;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.github.chenxiaolong.dualbootpatcher.R;
import com.github.chenxiaolong.dualbootpatcher.RomUtils.RomInformation;
import com.github.chenxiaolong.dualbootpatcher.switcher.RomDetailAdapter.BaseViewHolder;
import com.squareup.picasso.Picasso;

import java.io.File;
import java.util.List;

public class RomDetailAdapter extends RecyclerView.Adapter<BaseViewHolder> {
    public static final int ITEM_TYPE_ROM_CARD = 1;
    public static final int ITEM_TYPE_INFO = 2;
    public static final int ITEM_TYPE_ACTION = 3;

    public static class Item {
        int type;

        public Item(int type) {
            this.type = type;
        }
    }

    public static class RomCardItem extends Item {
        RomInformation romInfo;

        public RomCardItem(RomInformation romInfo) {
            super(ITEM_TYPE_ROM_CARD);
            this.romInfo = romInfo;
        }
    }

    public static class InfoItem extends Item {
        int id;
        String title;
        String value;

        public InfoItem(int id, String title, String value) {
            super(ITEM_TYPE_INFO);
            this.id = id;
            this.title = title;
            this.value = value;
        }
    }

    public static class ActionItem extends Item {
        int id;
        int iconResId;
        String title;

        public ActionItem(int id, int iconResId, String title) {
            super(ITEM_TYPE_ACTION);
            this.id = id;
            this.iconResId = iconResId;
            this.title = title;
        }
    }

    public static abstract class BaseViewHolder extends RecyclerView.ViewHolder {
        int id;

        BaseViewHolder(View itemView) {
            super(itemView);
        }

        public abstract void display(Item item);
    }

    public static class CardViewHolder extends BaseViewHolder {
        ImageView vThumbnail;
        TextView vName;
        TextView vVersion;
        TextView vBuild;

        CardViewHolder(View itemView) {
            super(itemView);
            vThumbnail = (ImageView) itemView.findViewById(R.id.rom_thumbnail);
            vName = (TextView) itemView.findViewById(R.id.rom_name);
            vVersion = (TextView) itemView.findViewById(R.id.rom_version);
            vBuild = (TextView) itemView.findViewById(R.id.rom_build);
        }

        @Override
        public void display(Item item) {
            RomCardItem romCardItem = (RomCardItem) item;
            RomInformation romInfo = romCardItem.romInfo;

            // Load thumbnail
            Context context = vThumbnail.getContext();
            File f = new File(romInfo.getThumbnailPath());
            if (f.exists() && f.canRead()) {
                Picasso.with(context).load(f).error(romInfo.getImageResId()).into(vThumbnail);
            } else {
                Picasso.with(context).load(romInfo.getImageResId()).into(vThumbnail);
            }

            // Load name, version, build
            vName.setText(romInfo.getName());
            vVersion.setText(romInfo.getVersion());
            vBuild.setText(romInfo.getBuild());
        }
    }

    public static class InfoViewHolder extends BaseViewHolder {
        TextView vTitle;
        TextView vValue;

        InfoViewHolder(View itemView) {
            super(itemView);
            vTitle = (TextView) itemView.findViewById(R.id.title);
            vValue = (TextView) itemView.findViewById(R.id.value);
        }

        @Override
        public void display(Item item) {
            InfoItem infoItem = (InfoItem) item;
            vTitle.setText(infoItem.title);
            vValue.setText(infoItem.value);
        }
    }

    public static class ActionViewHolder extends BaseViewHolder {
        private RomDetailItemClickListener mListener;

        ImageView vIcon;
        TextView vTitle;

        ActionViewHolder(View itemView, RomDetailItemClickListener listener) {
            super(itemView);
            mListener = listener;
            vIcon = (ImageView) itemView.findViewById(R.id.action_icon);
            vTitle = (TextView) itemView.findViewById(R.id.action_title);

            itemView.setOnClickListener(new OnClickListener() {
                @Override
                public void onClick(View view) {
                    mListener.onActionItemClicked(view, getAdapterPosition());
                }
            });
        }

        @Override
        public void display(Item item) {
            ActionItem actionItem = (ActionItem) item;
            vIcon.setImageResource(actionItem.iconResId);
            vTitle.setText(actionItem.title);
        }
    }

    private RomDetailItemClickListener mOnItemClicked = new RomDetailItemClickListener() {
        @Override
        public void onActionItemClicked(View view, int position) {
            if (mListener != null) {
                mListener.onActionItemSelected((ActionItem) mItems.get(position));
            }
        }
    };

    private interface RomDetailItemClickListener {
        void onActionItemClicked(View view, int position);
    }

    public interface RomDetailAdapterListener {
        void onActionItemSelected(ActionItem item);
    }

    private List<Item> mItems;
    private RomDetailAdapterListener mListener;

    RomDetailAdapter(List<Item> items, RomDetailAdapterListener listener) {
        mItems = items;
        mListener = listener;
    }

    @Override
    public int getItemViewType(int position) {
        return mItems.get(position).type;
    }

    @Override
    public BaseViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        int layoutId;

        switch (viewType) {
        case ITEM_TYPE_ROM_CARD:
            layoutId = R.layout.rom_detail_card_item;
            break;
        case ITEM_TYPE_INFO:
            layoutId = R.layout.rom_detail_info_item;
            break;
        case ITEM_TYPE_ACTION:
            layoutId = R.layout.rom_detail_action_item;
            break;
        default:
            throw new IllegalStateException("Invalid viewType ID");
        }

        View view = LayoutInflater
                .from(parent.getContext())
                .inflate(layoutId, parent, false);

        switch (viewType) {
        case ITEM_TYPE_ROM_CARD:
            return new CardViewHolder(view);
        case ITEM_TYPE_INFO:
            return new InfoViewHolder(view);
        case ITEM_TYPE_ACTION:
            return new ActionViewHolder(view, mOnItemClicked);
        default:
            throw new IllegalStateException("Invalid viewType ID");
        }
    }

    @Override
    public void onBindViewHolder(BaseViewHolder holder, int position) {
        holder.display(mItems.get(position));
    }

    @Override
    public int getItemCount() {
        return mItems.size();
    }

    public static class DividerItemDecoration extends RecyclerView.ItemDecoration {
        private static final int[] ATTRS = new int[]{ android.R.attr.listDivider };
        private Drawable mDivider;

        public DividerItemDecoration(Context context) {
            final TypedArray styledAttributes = context.obtainStyledAttributes(ATTRS);
            mDivider = styledAttributes.getDrawable(0);
            styledAttributes.recycle();
        }

        @Override
        public void onDraw(Canvas c, RecyclerView parent, RecyclerView.State state) {
            int left = parent.getPaddingLeft();
            int right = parent.getWidth() - parent.getPaddingRight();

            int childCount = parent.getChildCount();
            for (int i = 0; i < childCount; i++) {
                View child = parent.getChildAt(i);

                RecyclerView.LayoutParams params =
                        (RecyclerView.LayoutParams) child.getLayoutParams();

                int top = child.getBottom() + params.bottomMargin;
                int bottom = top + mDivider.getIntrinsicHeight();

                mDivider.setBounds(left, top, right, bottom);
                mDivider.draw(c);
            }
        }
    }
}

File: Android_GUI/src/com/github/chenxiaolong/dualbootpatcher/switcher/RomDetailAdapter.kt
/*
 * Copyright (C) 2015  Andrew Gunnerson <andrewgunnerson@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.github.chenxiaolong.dualbootpatcher.switcher

import android.content.Context
import android.graphics.Canvas
import android.graphics.drawable.Drawable
import android.support.annotation.ColorInt
import android.support.v4.content.ContextCompat
import android.support.v4.graphics.drawable.DrawableCompat
import android.support.v7.widget.RecyclerView
import android.util.TypedValue
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView

import com.github.chenxiaolong.dualbootpatcher.R
import com.github.chenxiaolong.dualbootpatcher.RomUtils.RomInformation
import com.github.chenxiaolong.dualbootpatcher.switcher.RomDetailAdapter.BaseViewHolder
import com.squareup.picasso.Picasso

import java.io.File

class RomDetailAdapter internal constructor(
        private val items: List<Item>,
        private val listener: RomDetailAdapterListener?
) : RecyclerView.Adapter<BaseViewHolder>() {
    private val onItemClicked = object : RomDetailItemClickListener {
        override fun onActionItemClicked(view: View, position: Int) {
            listener?.onActionItemSelected(items[position] as ActionItem)
        }
    }

    open class Item(internal var type: Int)

    class RomCardItem(internal var romInfo: RomInformation) : Item(ITEM_TYPE_ROM_CARD)

    class InfoItem(
            internal var id: Int,
            internal var title: String,
            internal var value: String
    ) : Item(ITEM_TYPE_INFO)

    class ActionItem(
            internal var id: Int,
            internal var iconResId: Int,
            internal var title: String
    ) : Item(ITEM_TYPE_ACTION)

    abstract class BaseViewHolder internal constructor(itemView: View)
            : RecyclerView.ViewHolder(itemView) {
        internal var id: Int = 0

        abstract fun display(item: Item)
    }

    class CardViewHolder internal constructor(itemView: View) : BaseViewHolder(itemView) {
        private var vThumbnail: ImageView = itemView.findViewById(R.id.rom_thumbnail)
        private var vName: TextView = itemView.findViewById(R.id.rom_name)
        private var vVersion: TextView = itemView.findViewById(R.id.rom_version)
        private var vBuild: TextView = itemView.findViewById(R.id.rom_build)

        override fun display(item: Item) {
            val romCardItem = item as RomCardItem
            val romInfo = romCardItem.romInfo

            // Load thumbnail
            val f = File(romInfo.thumbnailPath!!)
            if (f.exists() && f.canRead()) {
                Picasso.get().load(f).error(romInfo.imageResId).into(vThumbnail)
            } else {
                Picasso.get().load(romInfo.imageResId).into(vThumbnail)
            }

            // Load name, version, build
            vName.text = romInfo.name
            vVersion.text = romInfo.version
            vBuild.text = romInfo.build
        }
    }

    class InfoViewHolder internal constructor(itemView: View) : BaseViewHolder(itemView) {
        private var vTitle: TextView = itemView.findViewById(R.id.title)
        private var vValue: TextView = itemView.findViewById(R.id.value)

        override fun display(item: Item) {
            val infoItem = item as InfoItem
            vTitle.text = infoItem.title
            vValue.text = infoItem.value
        }
    }

    class ActionViewHolder internal constructor(
            itemView: View,
            private val listener: RomDetailItemClickListener
    ) : BaseViewHolder(itemView) {
        private var vIcon: ImageView = itemView.findViewById(R.id.action_icon)
        private var vTitle: TextView = itemView.findViewById(R.id.action_title)

        init {
            itemView.setOnClickListener {
                view -> listener.onActionItemClicked(view, adapterPosition)
            }
        }

        override fun display(item: Item) {
            val actionItem = item as ActionItem
            // Tint drawable
            val context = vIcon.context
            val drawable = ContextCompat.getDrawable(context, actionItem.iconResId)
            val wrapped = DrawableCompat.wrap(drawable!!)
            DrawableCompat.setTint(wrapped, getThemeTextColor(context))
            vIcon.setImageDrawable(wrapped)
            //vIcon.setImageResource(actionItem.iconResId);
            vTitle.text = actionItem.title
        }

        @ColorInt
        private fun getThemeTextColor(context: Context): Int {
            val value = TypedValue()
            context.theme.resolveAttribute(android.R.attr.textColor, value, true)
            return value.data
        }
    }

    internal interface RomDetailItemClickListener {
        fun onActionItemClicked(view: View, position: Int)
    }

    interface RomDetailAdapterListener {
        fun onActionItemSelected(item: ActionItem)
    }

    override fun getItemViewType(position: Int): Int {
        return items[position].type
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BaseViewHolder {
        val layoutId = when (viewType) {
            ITEM_TYPE_ROM_CARD -> R.layout.rom_detail_card_item
            ITEM_TYPE_INFO -> R.layout.rom_detail_info_item
            ITEM_TYPE_ACTION -> R.layout.rom_detail_action_item
            else -> throw IllegalStateException("Invalid viewType ID")
        }

        val view = LayoutInflater
                .from(parent.context)
                .inflate(layoutId, parent, false)

        return when (viewType) {
            ITEM_TYPE_ROM_CARD -> CardViewHolder(view)
            ITEM_TYPE_INFO -> InfoViewHolder(view)
            ITEM_TYPE_ACTION -> ActionViewHolder(view, onItemClicked)
            else -> throw IllegalStateException("Invalid viewType ID")
        }
    }

    override fun onBindViewHolder(holder: BaseViewHolder, position: Int) {
        holder.display(items[position])
    }

    override fun getItemCount(): Int {
        return items.size
    }

    class DividerItemDecoration(context: Context) : RecyclerView.ItemDecoration() {
        private val divider: Drawable?

        init {
            val styledAttributes = context.obtainStyledAttributes(ATTRS)
            divider = styledAttributes.getDrawable(0)
            styledAttributes.recycle()
        }

        override fun onDraw(c: Canvas, parent: RecyclerView, state: RecyclerView.State?) {
            val left = parent.paddingLeft
            val right = parent.width - parent.paddingRight

            val childCount = parent.childCount
            for (i in 0 until childCount) {
                val child = parent.getChildAt(i)

                val params = child.layoutParams as RecyclerView.LayoutParams

                val top = child.bottom + params.bottomMargin
                val bottom = top + divider!!.intrinsicHeight

                divider.setBounds(left, top, right, bottom)
                divider.draw(c)
            }
        }

        companion object {
            private val ATTRS = intArrayOf(android.R.attr.listDivider)
        }
    }

    companion object {
        const val ITEM_TYPE_ROM_CARD = 1
        const val ITEM_TYPE_INFO = 2
        const val ITEM_TYPE_ACTION = 3
    }
}